File: web/src/crypto.ts
// Copyright 2026 Marcelo Cantos
// SPDX-License-Identifier: Apache-2.0

const subtle = globalThis.crypto.subtle;

/** Generate 32 cryptographically random bytes suitable for use as a nonce. */
export function generateNonce(): Uint8Array {
  const b = new Uint8Array(32);
  crypto.getRandomValues(b);
  return b;
}

/** Generate 32 cryptographically random bytes suitable for use as a secret. */
export function generateSecret(): Uint8Array {
  return generateNonce();
}

/**
 * Derives a 256-bit key from input keying material via HKDF-SHA256.
 * Uses empty salt (matching Go's hkdf.New(sha256.New, ikm, nil, info)).
 */
export async function deriveKeyFromSecret(
  secret: Uint8Array,
  info: Uint8Array,
): Promise<Uint8Array> {
  const ikm = await subtle.importKey("raw", secret, "HKDF", false, [
    "deriveBits",
  ]);
  const bits = await subtle.deriveBits(
    {
      name: "HKDF",
      hash: "SHA-256",
      salt: new Uint8Array(0),
      info,
    },
    ikm,
    256,
  );
  return new Uint8Array(bits);
}

/**
 * Derives a 6-digit confirmation code from two ECDH public keys.
 * The code is order-independent: deriveConfirmationCode(a, b) === deriveConfirmationCode(b, a).
 */
export async function deriveConfirmationCode(
  pubA: Uint8Array,
  pubB: Uint8Array,
): Promise<string> {
  // Sort lexicographically
  let a = pubA;
  let b = pubB;
  if (compareBytes(a, b) > 0) {
    [a, b] = [b, a];
  }
  // Concat
  const ikm = new Uint8Array(a.length + b.length);
  ikm.set(a, 0);
  ikm.set(b, a.length);

  const derived = await deriveKeyFromSecret(
    ikm,
    new TextEncoder().encode("pairing-confirmation"),
  );
  // Take first 4 bytes big-endian, mod 1000000
  const view = new DataView(derived.buffer, derived.byteOffset, 4);
  const code = view.getUint32(0, false) % 1000000;
  return code.toString().padStart(6, "0");
}

function compareBytes(a: Uint8Array, b: Uint8Array): number {
  const len = Math.min(a.length, b.length);
  for (let i = 0; i < len; i++) {
    if (a[i] !== b[i]) return a[i] - b[i];
  }
  return a.length - b.length;
}

/**
 * X25519 ECDH key pair for key exchange.
 */
export class E2EKeyPair {
  private privateKey: CryptoKey;
  /** 32-byte raw X25519 public key. */
  public publicKeyData: Uint8Array;

  private constructor(privateKey: CryptoKey, publicKeyData: Uint8Array) {
    this.privateKey = privateKey;
    this.publicKeyData = publicKeyData;
  }

  /** Generate a new X25519 key pair. */
  static async create(): Promise<E2EKeyPair> {
    const keyPair = await subtle.generateKey("X25519", true, [
      "deriveBits",
    ]) as CryptoKeyPair;
    const rawPub = await subtle.exportKey("raw", keyPair.publicKey);
    return new E2EKeyPair(keyPair.privateKey, new Uint8Array(rawPub));
  }

  /** Export the raw private key bytes (32 bytes). */
  async exportPrivateKey(): Promise<Uint8Array> {
    // Node.js doesn't support "raw" export for X25519 private keys;
    // use PKCS#8 and strip the 16-byte ASN.1 header.
    const pkcs8 = await subtle.exportKey("pkcs8", this.privateKey);
    const bytes = new Uint8Array(pkcs8);
    // PKCS#8 for X25519: 16-byte header + 32-byte key
    return bytes.slice(bytes.length - 32);
  }

  /**
   * Perform ECDH with a peer's public key, then derive a 256-bit session
   * key via HKDF-SHA256.
   */
  async deriveSessionKey(
    peerPublicKey: Uint8Array,
    info: Uint8Array,
  ): Promise<Uint8Array> {
    const peerKey = await subtle.importKey(
      "raw",
      peerPublicKey,
      "X25519",
      false,
      [],
    );
    const sharedBits = await subtle.deriveBits(
      { name: "X25519", public: peerKey },
      this.privateKey,
      256,
    );
    const shared = new Uint8Array(sharedBits);
    return deriveKeyFromSecret(shared, info);
  }
}

/**
 * Controls how `E2EChannel.decrypt` handles sequence numbers.
 *
 * - `"strict"` (default): sequence numbers must be contiguous with no gaps.
 *   Any out-of-order or replayed packet is rejected. Suitable for reliable
 *   transports (WebTransport streams).
 * - `"datagrams"`: gaps in the sequence number space are accepted, as expected
 *   on lossy transports (UDP, H.264 video). Packets with a sequence number
 *   less than the last accepted one are rejected to prevent replay attacks.
 */
export type ChannelMode = "strict" | "datagrams";

/**
 * AES-256-GCM symmetric encryption channel with counter nonce.
 *
 * Wire format matches Go/Swift/Kotlin implementations:
 *   Nonce: 12 bytes — first 8 = sequence LE, last 4 = zero
 *   Ciphertext: [8-byte seq LE][ciphertext + 16-byte GCM tag]
 *   Additional data: the 8-byte sequence bytes
 */
export class E2EChannel {
  private sendKey: CryptoKey;
  private recvKey: CryptoKey;
  private sendSeq: bigint = 0n;
  private recvSeq: bigint = 0n;
  /** The sequence-number checking mode used by `decrypt`. Defaults to `"strict"`. */
  public mode: ChannelMode = "strict";

  private constructor(sendKey: CryptoKey, recvKey: CryptoKey) {
    this.sendKey = sendKey;
    this.recvKey = recvKey;
  }

  /** Create a channel from raw 32-byte send and receive keys. */
  static async create(
    sendKeyBytes: Uint8Array,
    recvKeyBytes: Uint8Array,
  ): Promise<E2EChannel> {
    const sendKey = await subtle.importKey(
      "raw",
      sendKeyBytes,
      "AES-GCM",
      false,
      ["encrypt"],
    );
    const recvKey = await subtle.importKey(
      "raw",
      recvKeyBytes,
      "AES-GCM",
      false,
      ["decrypt"],
    );
    return new E2EChannel(sendKey, recvKey);
  }

  /**
   * Create a channel from a shared key, deriving directional keys via HKDF.
   * Matches Go's NewSymmetricChannel.
   */
  static async fromSharedKey(
    sharedKey: Uint8Array,
    isServer: boolean,
  ): Promise<E2EChannel> {
    const enc = new TextEncoder();
    let sendInfo = enc.encode("client-to-server");
    let recvInfo = enc.encode("server-to-client");
    if (isServer) {
      [sendInfo, recvInfo] = [recvInfo, sendInfo];
    }
    const sendKeyBytes = await deriveKeyFromSecret(sharedKey, sendInfo);
    const recvKeyBytes = await deriveKeyFromSecret(sharedKey, recvInfo);
    return E2EChannel.create(sendKeyBytes, recvKeyBytes);
  }

  /** Encrypt plaintext. Returns [8-byte seq LE][ciphertext + 16-byte tag]. */
  async encrypt(plaintext: Uint8Array): Promise<Uint8Array> {
    const seq = this.sendSeq;
    this.sendSeq++;

    const seqBytes = new Uint8Array(8);
    new DataView(seqBytes.buffer).setBigUint64(0, seq, true); // LE

    const nonce = new Uint8Array(12);
    nonce.set(seqBytes, 0); // first 8 bytes = seq LE, last 4 = zero

    const ciphertext = await subtle.encrypt(
      { name: "AES-GCM", iv: nonce, additionalData: seqBytes },
      this.sendKey,
      plaintext,
    );

    // Prepend seq bytes
    const result = new Uint8Array(8 + ciphertext.byteLength);
    result.set(seqBytes, 0);
    result.set(new Uint8Array(ciphertext), 8);
    return result;
  }

  /** Decrypt data. Verifies the sequence number according to `mode`. */
  async decrypt(data: Uint8Array): Promise<Uint8Array> {
    if (data.length < 8) {
      throw new Error("ciphertext too short");
    }

    const seqBytes = data.slice(0, 8);
    const seq = new DataView(
      seqBytes.buffer,
      seqBytes.byteOffset,
    ).getBigUint64(0, true);
    const ciphertext = data.slice(8);

    if (this.mode === "strict") {
      if (seq !== this.recvSeq) {
        throw new Error("unexpected sequence number");
      }
      this.recvSeq++;
    } else {
      // datagrams: recvSeq holds highest-accepted-seq + 1 (0n if none yet).
      // Accept seq >= recvSeq (gaps allowed); reject seq < recvSeq (replay).
      if (seq < this.recvSeq) {
        throw new Error("sequence number replayed or too old");
      }
      this.recvSeq = seq + 1n;
    }

    const nonce = new Uint8Array(12);
    nonce.set(seqBytes, 0);

    const plaintext = await subtle.decrypt(
      { name: "AES-GCM", iv: nonce, additionalData: seqBytes },
      this.recvKey,
      ciphertext,
    );
    return new Uint8Array(plaintext);
  }
}

// ---- Pairing record ----

/** Persistent state from a completed pairing ceremony. */
export interface PairingRecord {
  peerInstanceID: string;
  relayURL: string;
  localPrivateKey: string; // base64-encoded raw X25519 private key
  localPublicKey: string; // base64-encoded raw X25519 public key
  peerPublicKey: string; // base64-encoded raw X25519 public key
}

/** Create a pairing record from a completed ceremony. */
export async function createPairingRecord(
  peerInstanceID: string,
  relayURL: string,
  localKeyPair: E2EKeyPair,
  peerPublicKey: Uint8Array,
): Promise<PairingRecord> {
  const privBytes = await localKeyPair.exportPrivateKey();
  return {
    peerInstanceID,
    relayURL,
    localPrivateKey: btoa(String.fromCharCode(...privBytes)),
    localPublicKey: btoa(String.fromCharCode(...localKeyPair.publicKeyData)),
    peerPublicKey: btoa(String.fromCharCode(...peerPublicKey)),
  };
}

/** Derive a channel from a stored pairing record. */
export async function deriveChannelFromRecord(
  record: PairingRecord,
  sendInfo: Uint8Array,
  recvInfo: Uint8Array,
): Promise<E2EChannel> {
  const privateKeyBytes = Uint8Array.from(atob(record.localPrivateKey), (c) =>
    c.charCodeAt(0),
  );
  const peerPubBytes = Uint8Array.from(atob(record.peerPublicKey), (c) =>
    c.charCodeAt(0),
  );

  // PKCS#8 header for X25519 private keys (16 bytes).
  const pkcs8Header = new Uint8Array([
    0x30, 0x2e, 0x02, 0x01, 0x00, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x6e,
    0x04, 0x22, 0x04, 0x20,
  ]);
  const pkcs8 = new Uint8Array(pkcs8Header.length + privateKeyBytes.length);
  pkcs8.set(pkcs8Header, 0);
  pkcs8.set(privateKeyBytes, pkcs8Header.length);

  // Import private key via PKCS#8 (Node.js doesn't support raw X25519 import).
  const privateKey = await subtle.importKey(
    "pkcs8",
    pkcs8,
    "X25519",
    false,
    ["deriveBits"],
  );
  const peerPublicKey = await subtle.importKey(
    "raw",
    peerPubBytes,
    "X25519",
    false,
    [],
  );

  // Derive shared secret via ECDH.
  const sharedBits = await subtle.deriveBits(
    { name: "X25519", public: peerPublicKey },
    privateKey,
    256,
  );
  const shared = new Uint8Array(sharedBits);

  // Derive directional keys via HKDF.
  const sendKey = await deriveKeyFromSecret(shared, sendInfo);
  const recvKey = await deriveKeyFromSecret(shared, recvInfo);

  return E2EChannel.create(sendKey, recvKey);
}

File: crypto/crypto.go
// Copyright 2026 Marcelo Cantos
// SPDX-License-Identifier: Apache-2.0

// Package crypto provides end-to-end encryption for WebSocket traffic
// relayed through tern. The relay sees only ciphertext.
//
// Key exchange uses ECDH (X25519). Symmetric encryption uses AES-256-GCM
// with a monotonic counter nonce. Session keys are derived via HKDF-SHA256.
package crypto

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/ecdh"
	"crypto/rand"
	"crypto/sha256"
	"encoding/binary"
	"errors"
	"fmt"
	"sync"
	"sync/atomic"

	"golang.org/x/crypto/hkdf"
)

// KeyPair holds an ECDH X25519 key pair for key exchange.
type KeyPair struct {
	Private *ecdh.PrivateKey
	Public  *ecdh.PublicKey
}

// GenerateKeyPair creates a new X25519 key pair.
func GenerateKeyPair() (*KeyPair, error) {
	priv, err := ecdh.X25519().GenerateKey(rand.Reader)
	if err != nil {
		return nil, err
	}
	return &KeyPair{Private: priv, Public: priv.PublicKey()}, nil
}

// DeriveSessionKey performs ECDH with the peer's public key and derives
// a 256-bit AES key via HKDF-SHA256.
func DeriveSessionKey(priv *ecdh.PrivateKey, peerPub *ecdh.PublicKey, info []byte) ([]byte, error) {
	shared, err := priv.ECDH(peerPub)
	if err != nil {
		return nil, err
	}
	return hkdfDerive(shared, info)
}

// DeriveKeyFromSecret derives a session key from a persistent secret
// and a random nonce via HKDF-SHA256.
func DeriveKeyFromSecret(secret, nonce []byte) ([]byte, error) {
	return hkdfDerive(secret, nonce)
}

func hkdfDerive(ikm, info []byte) ([]byte, error) {
	r := hkdf.New(sha256.New, ikm, nil, info)
	key := make([]byte, 32)
	if _, err := r.Read(key); err != nil {
		return nil, err
	}
	return key, nil
}

// GenerateNonce creates a random 32-byte nonce for session key derivation.
func GenerateNonce() ([]byte, error) {
	nonce := make([]byte, 32)
	_, err := rand.Read(nonce)
	return nonce, err
}

// GenerateSecret creates a random 32-byte persistent device secret.
func GenerateSecret() ([]byte, error) {
	secret := make([]byte, 32)
	_, err := rand.Read(secret)
	return secret, err
}

// DeriveConfirmationCode computes a 6-digit confirmation code from
// both ECDH public keys. The code is deterministic and
// order-independent: DeriveConfirmationCode(a, b) == DeriveConfirmationCode(b, a).
//
// Both sides of the key exchange compute this independently. Under
// honest conditions, both derive the same code. Under a MitM attack
// (where the adversary substituted its own public key), each side
// computes a different code — the mismatch aborts pairing.
func DeriveConfirmationCode(pubA, pubB *ecdh.PublicKey) (string, error) {
	a, b := pubA.Bytes(), pubB.Bytes()
	if bytes.Compare(a, b) > 0 {
		a, b = b, a
	}
	ikm := append(a, b...)
	r := hkdf.New(sha256.New, ikm, nil, []byte("pairing-confirmation"))
	buf := make([]byte, 4)
	if _, err := r.Read(buf); err != nil {
		return "", err
	}
	code := binary.BigEndian.Uint32(buf) % 1000000
	return fmt.Sprintf("%06d", code), nil
}

// ChannelMode controls how Decrypt handles sequence numbers.
type ChannelMode int

const (
	// ModeStrict (default) requires sequence numbers to be strictly
	// monotonic with no gaps. Any out-of-order or replayed packet is
	// rejected. Suitable for reliable transports (TCP/WebSocket).
	ModeStrict ChannelMode = iota

	// ModeDatagrams allows gaps in the sequence number space, which
	// is expected on lossy transports (UDP, H.264 video streams).
	// Any packet with a sequence number greater than the last received
	// is accepted and the counter jumps forward. Packets with a
	// sequence number less than or equal to the last received are
	// rejected to prevent replay attacks.
	ModeDatagrams
)

// Channel provides symmetric encryption/decryption for a WebSocket
// connection. Uses AES-256-GCM with a monotonic counter nonce to
// prevent nonce reuse.
type Channel struct {
	sendGCM cipher.AEAD
	recvGCM cipher.AEAD
	sendSeq atomic.Uint64
	recvSeq uint64
	recvMu  sync.Mutex
	mode    ChannelMode
}

// NewChannel creates an encrypted channel from a shared key.
// sendKey and recvKey should be different (derived with different info
// strings) to prevent nonce collision on the two directions.
func NewChannel(sendKey, recvKey []byte) (*Channel, error) {
	sendBlock, err := aes.NewCipher(sendKey)
	if err != nil {
		return nil, err
	}
	sendGCM, err := cipher.NewGCM(sendBlock)
	if err != nil {
		return nil, err
	}

	recvBlock, err := aes.NewCipher(recvKey)
	if err != nil {
		return nil, err
	}
	recvGCM, err := cipher.NewGCM(recvBlock)
	if err != nil {
		return nil, err
	}

	return &Channel{sendGCM: sendGCM, recvGCM: recvGCM}, nil
}

// NewDatagramChannel creates a channel in ModeDatagrams. It is a
// convenience wrapper around NewChannel for lossy transports such as
// UDP or H.264 video streams.
func NewDatagramChannel(sendKey, recvKey []byte) (*Channel, error) {
	ch, err := NewChannel(sendKey, recvKey)
	if err != nil {
		return nil, err
	}
	ch.mode = ModeDatagrams
	return ch, nil
}

// SetMode changes the sequence-number checking behaviour of Decrypt.
// It is safe to call before any Decrypt calls; calling it concurrently
// with Decrypt is not supported.
func (c *Channel) SetMode(mode ChannelMode) {
	c.recvMu.Lock()
	defer c.recvMu.Unlock()
	c.mode = mode
}

// NewSymmetricChannel creates a channel where both directions use the
// same key but different nonce prefixes (0x00 for client→server,
// 0x01 for server→client). Use this for the persistent secret path.
func NewSymmetricChannel(key []byte, isServer bool) (*Channel, error) {
	sendInfo := []byte("client-to-server")
	recvInfo := []byte("server-to-client")
	if isServer {
		sendInfo, recvInfo = recvInfo, sendInfo
	}

	sendKey, err := hkdfDerive(key, sendInfo)
	if err != nil {
		return nil, err
	}
	recvKey, err := hkdfDerive(key, recvInfo)
	if err != nil {
		return nil, err
	}

	return NewChannel(sendKey, recvKey)
}

// Encrypt encrypts a plaintext message. The returned ciphertext includes
// the 8-byte sequence number prefix.
func (c *Channel) Encrypt(plaintext []byte) []byte {
	seq := c.sendSeq.Add(1) - 1
	nonce := makeNonce(c.sendGCM.NonceSize(), seq)
	seqBytes := make([]byte, 8)
	binary.LittleEndian.PutUint64(seqBytes, seq)
	ciphertext := c.sendGCM.Seal(nil, nonce, plaintext, seqBytes)
	return append(seqBytes, ciphertext...)
}

// Decrypt decrypts a ciphertext message. Verifies the sequence number
// to prevent replay attacks.
func (c *Channel) Decrypt(data []byte) ([]byte, error) {
	if len(data) < 8 {
		return nil, errors.New("ciphertext too short")
	}

	seqBytes := data[:8]
	seq := binary.LittleEndian.Uint64(seqBytes)
	ciphertext := data[8:]

	c.recvMu.Lock()
	defer c.recvMu.Unlock()

	switch c.mode {
	case ModeStrict:
		if seq != c.recvSeq {
			return nil, errors.New("unexpected sequence number")
		}
		c.recvSeq++
	case ModeDatagrams:
		// c.recvSeq holds the highest accepted seq + 1 (or 0 if none yet).
		// Accept any seq >= c.recvSeq (gaps allowed); reject seq < c.recvSeq
		// (replay protection — we have already received something at or after
		// this position).
		if seq < c.recvSeq {
			return nil, errors.New("sequence number replayed or too old")
		}
		c.recvSeq = seq + 1
	}

	nonce := makeNonce(c.recvGCM.NonceSize(), seq)
	return c.recvGCM.Open(nil, nonce, ciphertext, seqBytes)
}

func makeNonce(size int, seq uint64) []byte {
	nonce := make([]byte, size)
	binary.LittleEndian.PutUint64(nonce, seq)
	return nonce
}
